add DroneerF405 hwdef

DroneerUAV opened the pull request:
Add DroneerF405 board from Droneer UAV

andyp1per commented:
I recommend these changes:
```
diff --git a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
index 4063f183a1..d45d1aec11 100644
--- a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
+++ b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
@@ -34,8 +34,8 @@ PA12 OTG_FS_DP OTG1
 PC5 VBUS INPUT OPENDRAIN

 # USART1(VTX)
-PA9  USART1_TX USART1
-PA10 USART1_RX USART1
+PA9  USART1_TX USART1 NODMA
+PA10 USART1_RX USART1 NODMA
 define DEFAULT_SERIAL1_PROTOCOL SerialProtocol_Tramp

 # USART2, RC input
@@ -70,10 +70,10 @@ define DEFAULT_SERIAL5_PROTOCOL SerialProtocol_None

 # PWM output. 1 - 8 on ESC header
 # --------------------- PWM -----------------------
-PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50)
+PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50) BIDIR
 PC7  TIM8_CH2   TIM8  PWM(2)  GPIO(51)
 PC8  TIM8_CH3   TIM8  PWM(3)  GPIO(52)
-PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53)
+PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53) BIDIR
 PA15 TIM2_CH1   TIM2  PWM(5)  GPIO(54)
 PA8  TIM1_CH1   TIM1  PWM(6)  GPIO(55)
 PB10 TIM2_CH3   TIM2  PWM(7)  GPIO(56)
@@ -174,7 +174,7 @@ define HAL_OSD_TYPE_DEFAULT 1
 # Font for the osd
 ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin

-DMA_PRIORITY TIM2* TIM3*
+DMA_PRIORITY TIM8* TIM3*
```

andyp1per reviewed (changes requested):
Needs a README.md
Suggest the changes I outlined

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef-bl.dat:
```diff
@@ -0,0 +1,39 @@
+OSCILLATOR_HZ 8000000
+
+# board ID for firmware load
+APJ_BOARD_ID 1054
```
This ID is used. Please reserve an unused id

peterbarker replied:
@DroneerUAV to be clear, you need to make an entry in `Tools/AP_Bootloader/board_types.txt`, reserving an ID for your board.  You then you use either that numeric ID in here, or the board name you've used in `board_types.txt`

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat:
```diff
@@ -0,0 +1,182 @@
+HAL_CHIBIOS_ARCH_F405 1
+
+# board ID for firmware load
+APJ_BOARD_ID 1054
```
This id is used - need another one

DroneerUAV commented:
> 我建议进行以下更改：
> 
> ```
> diff --git a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> index 4063f183a1..96d622b26a 100644
> --- a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> +++ b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> @@ -34,14 +34,14 @@ PA12 OTG_FS_DP OTG1
>  PC5 VBUS INPUT OPENDRAIN
> 
>  # USART1(VTX)
> -PA9  USART1_TX USART1
> -PA10 USART1_RX USART1
> +PA9  USART1_TX USART1 NODMA
> +PA10 USART1_RX USART1 NODMA
>  define DEFAULT_SERIAL1_PROTOCOL SerialProtocol_Tramp
> 
>  # USART2, RC input
>  PA2 USART2_TX USART2
> -PA3 USART2_RX USART2 NODMA
> -define DEFAULT_SERIAL2_PROTOCOL SerialProtocol_MAVLink2
> +PA3 USART2_RX USART2
> +define DEFAULT_SERIAL2_PROTOCOL SerialProtocol_RCIN
> 
>  # default to timer for RC input
>  #PA3 TIM9_CH2 TIM9 RCININT PULLDOWN LOW
> @@ -70,10 +70,10 @@ define DEFAULT_SERIAL5_PROTOCOL SerialProtocol_None
> 
>  # PWM output. 1 - 8 on ESC header
>  # --------------------- PWM -----------------------
> -PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50)
> +PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50) BIDIR
>  PC7  TIM8_CH2   TIM8  PWM(2)  GPIO(51)
>  PC8  TIM8_CH3   TIM8  PWM(3)  GPIO(52)
> -PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53)
> +PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53) BIDIR
>  PA15 TIM2_CH1   TIM2  PWM(5)  GPIO(54)
>  PA8  TIM1_CH1   TIM1  PWM(6)  GPIO(55)
>  PB10 TIM2_CH3   TIM2  PWM(7)  GPIO(56)
> @@ -174,7 +174,7 @@ define HAL_OSD_TYPE_DEFAULT 1
>  # Font for the osd
>  ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin
> 
> -DMA_PRIORITY TIM2* TIM3*
> +DMA_PRIORITY TIM8* TIM3*
> 
>  # minimal drivers to reduce flash usage
>  include ../include/minimize_fpv_osd.inc
> ```

![1734343967586(1)](https://github.com/user-attachments/assets/1591dd1a-d686-43ce-bded-60389341a8e0)
Does the suggestion here need to be changed? Our hardware serial port 2 must be added with NODMA in order to work normally

andyp1per commented:
> > 我建议进行以下更改：
> > ```
> > diff --git a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> > index 4063f183a1..96d622b26a 100644
> > --- a/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> > +++ b/libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/hwdef.dat
> > @@ -34,14 +34,14 @@ PA12 OTG_FS_DP OTG1
> >  PC5 VBUS INPUT OPENDRAIN
> > 
> >  # USART1(VTX)
> > -PA9  USART1_TX USART1
> > -PA10 USART1_RX USART1
> > +PA9  USART1_TX USART1 NODMA
> > +PA10 USART1_RX USART1 NODMA
> >  define DEFAULT_SERIAL1_PROTOCOL SerialProtocol_Tramp
> > 
> >  # USART2, RC input
> >  PA2 USART2_TX USART2
> > -PA3 USART2_RX USART2 NODMA
> > -define DEFAULT_SERIAL2_PROTOCOL SerialProtocol_MAVLink2
> > +PA3 USART2_RX USART2
> > +define DEFAULT_SERIAL2_PROTOCOL SerialProtocol_RCIN
> > 
> >  # default to timer for RC input
> >  #PA3 TIM9_CH2 TIM9 RCININT PULLDOWN LOW
> > @@ -70,10 +70,10 @@ define DEFAULT_SERIAL5_PROTOCOL SerialProtocol_None
> > 
> >  # PWM output. 1 - 8 on ESC header
> >  # --------------------- PWM -----------------------
> > -PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50)
> > +PC6  TIM8_CH1   TIM8  PWM(1)  GPIO(50) BIDIR
> >  PC7  TIM8_CH2   TIM8  PWM(2)  GPIO(51)
> >  PC8  TIM8_CH3   TIM8  PWM(3)  GPIO(52)
> > -PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53)
> > +PC9  TIM8_CH4   TIM8  PWM(4)  GPIO(53) BIDIR
> >  PA15 TIM2_CH1   TIM2  PWM(5)  GPIO(54)
> >  PA8  TIM1_CH1   TIM1  PWM(6)  GPIO(55)
> >  PB10 TIM2_CH3   TIM2  PWM(7)  GPIO(56)
> > @@ -174,7 +174,7 @@ define HAL_OSD_TYPE_DEFAULT 1
> >  # Font for the osd
> >  ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin
> > 
> > -DMA_PRIORITY TIM2* TIM3*
> > +DMA_PRIORITY TIM8* TIM3*
> > 
> >  # minimal drivers to reduce flash usage
> >  include ../include/minimize_fpv_osd.inc
> > ```
> 
> ![1734343967586(1)](https://private-user-images.githubusercontent.com/191452273/396053005-1591dd1a-d686-43ce-bded-60389341a8e0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1NTgxNTksIm5iZiI6MTczNTU1Nzg1OSwicGF0aCI6Ii8xOTE0NTIyNzMvMzk2MDUzMDA1LTE1OTFkZDFhLWQ2ODYtNDNjZS1iZGVkLTYwMzg5MzQxYThlMC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIzMFQxMTI0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZGM3MDlmODhkMDliNmYxMTNlN2I3Nzc1ODgyNWMyZmExMmZlZTYxM2JkYjgxNTE5ZjJlM2IwNzE2OTk1YjRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9KqP_14Ol6s-dgJqUy2AcxQa2pC8ktITbgmzeqljasM) Does the suggestion here need to be changed? Our hardware serial port 2 must be added with NODMA in order to work normally

I don't think so. We discussed this on discord, you need to use an actual serial protocol

DroneerUAV commented:
@andyp1per I have addressed the requested changes,Please review the updated PR

andyp1per commented on Tools/AP_Bootloader/board_types.txt:
```diff
@@ -413,6 +413,13 @@ AP_HW_ZeroOne_GNSS                   5602
 AP_HW_DroneBuild_G2                  5701
 
+#IDs 5700-5710 reserved for DroneBuild  
+AP_HW_DroneBuild_G1                  5700
```
This is a dup of the two lines above please remove

DroneerUAV replied:
oh sry,now that I've changed it,please review

andyp1per reviewed (changes requested):
Looking good

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+## Features
+
+ - STM32F405 microcontroller
+ - MPU6000 IMU
```
incorrect
```suggestion
 - ICM42688-P IMU
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+ - MPU6000 IMU
+ - max7456 OSD
+ - 5 UARTs
+ - 9 PWM outputs
```
add other features
```suggestion
 - 9 PWM outputs
 - Dual Camera inputs
 - 3S to 8S LIPO operation
 - 5V,3A BEC, 10V,2A VTX BEC
 - Physical: 37x37mm, 7g
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+receive pin for UARTn. The Tn pin is the transmit pin for UARTn.
+
+ - SERIAL0 -> USB
+ - SERIAL1 -> UART1 (Tramp, DMA-enabled)
```
```suggestion
 - SERIAL1 -> UART1 (Tramp)
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+ - SERIAL0 -> USB
+ - SERIAL1 -> UART1 (Tramp, DMA-enabled)
+ - SERIAL2 -> UART2 (RCIN, DMA-enabled) 
+ - SERIAL3 -> UART3 (ESC)
```
```suggestion
 - SERIAL3 -> UART3 (User)
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+ - SERIAL2 -> UART2 (RCIN, DMA-enabled) 
+ - SERIAL3 -> UART3 (ESC)
+ - SERIAL4 -> UART4 (GPS)
+ - SERIAL5 -> UART5 (None)
```
```suggestion
 - SERIAL5 -> UART5 (ESC Telem, on ESC connector)
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+ - SERIAL5 -> UART5 (None)
+
+## RC Input
+
```
Replace the RC section with this:
```suggestion
The SBUS pin (SBU), is passed by an inverter to RX2 (UART2 RX), which can be used for all ArduPilot supported receiver protocols, except CRSF/ELRS and SRXL2. CRSF/ELRS and SRXL2 require the RX2 and TX2 pin be connected instead of the SBUS pin.

* SRXL2 requires SERIAL2_OPTIONS be set to “4”.
* Fport should be connected to TX2 via a bi-directional inverter and SERIAL2_OPTIONS be set to “4”.

Any UART can also be used for RC system connections in ArduPilot and is compatible with all protocols except PPM (SBUS requires external inversion on other UARTs). See Radio Control Systems for details.
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+  
+## OSD Support
+
+The Droneer F405 supports OSD using OSD_TYPE 1 (MAX7456 driver).
```
```suggestion
The Droneer F405 supports analog OSD using OSD_TYPE 1 (MAX7456 driver). Simultaneous HD TVX OSD operation can also be obtained using an available UART whose protocol is set to "DisplayPort (42)" and OSD_TYPE2 set to "5".
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+
+The Droneer F405 supports OSD using OSD_TYPE 1 (MAX7456 driver).
+
+## VTX Support
```
remove this secttion...no such connector exists

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+## PWM Output
+
+The Droneer F405 supports up to 9 PWM outputs. The pads for motor output
+M1 to M4 on the motor connector, plus M9 for LED strip or another
```
```suggestion
M1 to M4 on the motor connector, plus M9(marked LED) for LED strip or another
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+## Battery Monitoring
+
+The board has a internal voltage sensor and connections on the ESC connector for an external current sensor input.
+The voltage sensor can handle up to 6S.
```
```suggestion
The voltage sensor can handle up to 8S.
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+## Compass
+
+The Droneer F405 does not have a builtin compass, but you can attach an external compass using I2C on the SDA and SCL pads.
+
```
add this section:
```suggestion
## Firmware
Firmware for this board can be found here in sub-folders labeled “DroneerF405i”.

```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/DroneerF405/README.md:
```diff
@@ -0,0 +1,95 @@
+
+Once the initial firmware is loaded you can update the firmware using
+any ArduPilot ground station software. Updates should be done with the
+*.apj firmware files.
```
```suggestion
\*.apj firmware files.
```

Hwurzburg reviewed (changes requested):
in addition to changes listed, commits will need to be squashed and then split into one for Tools: add DronnerF405(for Tools library files)
and hwdef: add DroneerF405(for hwdef library files)

DroneerUAV commented:
> in addition to changes listed, commits will need to be squashed and then split into one for Tools: add DronnerF405(for Tools library files) and hwdef: add DroneerF405(for hwdef library files)

I've made all the required changes, please review

andyp1per commented:
Apologies, the commit messages need to match the module names so:
Tools -> bootloaders
hwdef->AP_HAL_ChibiOS

andyp1per commented on Tools/AP_Bootloader/board_types.txt:
```diff
@@ -413,6 +413,9 @@ AP_HW_ZeroOne_GNSS                   5602
 AP_HW_DroneBuild_G2                  5701
 
+#IDs 5800-5899 reserved for Droneer
+AP_HW_DroneerF405                    5800
```
We are not giving out 100 ids any more, please can you pick a smaller range of 10-20

peterbarker replied:
Just 10, please.  You can always come back for more later.

DroneerUAV replied:
got it

andyp1per reviewed (changes requested):
Some changes needed

DroneerUAV commented:
> Apologies, the commit messages need to match the module names so: Tools -> bootloaders hwdef->AP_HAL_ChibiOS

the commit message,like this?
Tools -> bootloaders 
AP_HAL_ChibiOS->hwdef

andyp1per commented:
Like this:
```
bootloaderss: add DroneerF405
AP_HAL_ChibiOS: add DroneerF405
```

DroneerUAV commented:
> Like this:
> 
> ```
> bootloaders: add DroneerF405
> AP_HAL_ChibiOS: add DroneerF405
> ```
> 
> and you actually need to split the bootloaders commit into two since the board id is in a separate module:
> 
> ```
> AP_Bootloader: add DroneerF405
> ```

ok,got it

DroneerUAV commented:
> Like this:
> 
> ```
> bootloaders: add DroneerF405
> AP_HAL_ChibiOS: add DroneerF405
> ```
> 
> and you actually need to split the bootloaders commit into two since the board id is in a separate module:
> 
> ```
> AP_Bootloader: add DroneerF405
> ```

I've made all the required changes, please review

Hwurzburg commented:
@DroneerUAV I have made a few changes that you had missed and cleaned up the RC section in the readme, rebased, and pushed here....should be good to go now

DroneerUAV commented:
> @DroneerUAV I have made a few changes that you had missed and cleaned up the RC section in the readme, rebased, and pushed here....should be good to go now

ok, thank you,Is there anything else I need to do

Hwurzburg commented:
I dont think so.. @andyp1per can you approve?

andyp1per reviewed (approved):
LGTM, needs a rebase

DroneerUAV commented:
> LGTM, needs a rebase

I have rebase to the latest

peterbarker commented:
Merged this, thanks @DroneerUAV .  Please note in future you might want to create a topic branch to create your pull request from; using your master branch for this sort of thing isn't great :-)  See https://ardupilot.org/dev/docs/git-branch.html

DroneerUAV commented:
> Merged this, thanks @DroneerUAV . Please note in future you might want to create a topic branch to create your pull request from; using your master branch for this sort of thing isn't great :-) See https://ardupilot.org/dev/docs/git-branch.html

ok,got it,thanks
